harvester/tests
# Add initial Cypress Rancher Integration test cases

### Purpose 
Develop Cypress test for Harvester Rancher integration related test cases. 
Plan to fulfill the requirement to automated basic regression and KDM release test.  

### Development Environment 
The development based on the following SUT version and UI

* Rancher version: v2.6.7 
* Harvester version: v1.0.3 
* Cypress version: 10.2 

### Prepare Test Environment 
We can prepare the Rancher integration test environment in the following manner  to provision Harvester with Rancher, 
1. Use vagrant ipxe-example pull request [45](https://github.com/harvester/ipxe-examples/pull/45), set
  - harvester_network_config.offline: `false` 
  - rancher_config.run_single_node_rancher: `true`  
  - rancher_config.run_single_node_air_gapped_rancher: `false`

2. Create separate VMs, one for Harvester and another for Rancher, make sure the network connectivity
3. Install Harvester and Rancher on bare machines and ensure the connectivity

### About Rancher First Time Login
_**Please be noted**_
In every test, while login Rancher, it will automatically check whether it's the first time login 
The test function `rancher.rancherLogin()` support bootstrap login in the following scenarios.

1. Use vagrant ipxe-example to prepare Harvester and Rancher and install `sshpass` on your local test runner environment 
2. You can also use regular VM or bare machine while you already setup the following 
 - ssh root permission to Rancher host VM or bare machine 
 - Install `sshpass` on your local test runner environment 

**_If your environment did not fulfill above requirement, we would suggest you finishing the Rancher bootstrap on your own before running the rancher integration test_**

### Cypress setup
1. Execute `npm install`
3. Execute `npm ci` 
4. Rename `cypress.env.json.example` to `cypress.env.json` file, update and add the following content 
  ```
  "username": "admin",
  "password": "passwdpasswd",
  "baseUrl": "https://192.168.122.149", 
  "rancherUrl": "https://192.168.122.224",
  "rancher_user": "admin",
  "rancher_password": "passwdpasswd",
  "rancher_vm_user": "vagrant",  
  "rancher_vm_password": "vagrant",   
   "vagrant_pxe_path": "../ipxe-examples/vagrant-pxe-harvester",
  ```

### Check test data
You can check and modify all related rancher integration test data in `../cypress/fixtures/rancher.json`
The test data here is suitable for `Rancher v2.6.7` along with the current KDM release version 
Please be noted, on new KDM release, you may need to update these value accordingly. 

```
    "rke2_latest": "1.24.4",
    "rke2_stable": "1.23.10",
    "k3s_latest": "1.24.4",
    "k3s_stable": "1.23.10", 
    "rke1_latest": "1.24.4",
    "rke1_stable": "1.23.10",
```
And also need to check the image and ssh user name to fulfill what you used in the `Prepare Harvester Image` and `Prepare Harvester VLAN network` test case, here I use focal-server and default vlan1 

```
    "image": "focal-server",
    "network_name": "default/vlan1",
    "ssh_user": "ubuntu",
```

I use ubuntu 20.04 cloud focal by settings the image name and url in  `cypress.env.json`

```
"image":
  {
    "name": "openSUSE-Leap-15.3-3-DVD-x86_64-Build38.1-Media.iso",
    "url": "https://download.opensuse.org/distribution/leap/15.3/iso/openSUSE-Leap-15.3-3-DVD-x86_64-Build38.1-Media.iso"
  }
```


### Execute Test case
Execute `npx cypress open` or `npx cypress run` for the test 

###  Test case list 
1. Prepare Harvester Image (On Harvester)
2. Prepare Harvester VLAN network (On Harvester)
3. Rancher import Harvester 
4. Harvester import Rancher (On Harvester)
5. Check Harvester Cluster Status (On Rancher) 
6. Create Cloud Credential
7. Provisiong RKE2 Cluster
8. Check RKE2 Cluster Status
9. Provisiong K3s Cluster
10. Check K3s Cluster Status
11. Verify RKE2 CSI driver
12. Verify RKE2 Cloud Provider
13. Verify RKE2 DHDP Load Balancer
14. Create Harvester VIP Pool (On Harvester)
15. Create RKE2 Pool Load Balancer
16. Delete Active RKE2 Cluster
17. Delete Active K3s Cluster
18. Delete Cloud Credential
19. Delete imported Harveter cluster

###  Recommended group of tests 
1. Complete test flow from test 1 to 19, can fulfill from import Harvester, create credential, provisioning RKE2 and K3s cluster, CSI driver, cloud provider and finally cleanup environment to ensure can rerun every time.
2. Run test 1 to 6 to prepare an imported Harvester ready for more test 
20. Run test 7, 8, 11, 12, 13 ,14, 15, 16, 18, 19, to verify cloud provider and csi driver on RKE2 cluster and cleanup environment 
21. Run cleanup from test 16 to 19 to cleanup the current the environment  

### Validation before commit 

- Rancher import Harvester 
- Harvester import Rancher (On Harvester)
- Check Harvester Cluster Status (On Rancher) 
- Create Cloud Credential

  ![image](https://user-images.githubusercontent.com/29251855/188810163-48ecd793-fc7c-4488-b89e-cf48e6680958.png)

- Provisiong RKE2 Cluster
- Check RKE2 Cluster Status
- Verify RKE2 CSI driver
- Verify RKE2 Cloud Provider
- Verify RKE2 DHDP Load Balancer
- Create Harvester VIP Pool
- Create RKE2 Pool Load Balancer

  ![image](https://user-images.githubusercontent.com/29251855/188813536-11e013cf-bcab-4a6e-b466-d8c63db89ab1.png)

- Delete Active RKE2 Cluster
- Delete Cloud Credential
- Delete imported Harvester cluster

  ![image](https://user-images.githubusercontent.com/29251855/188830295-d95749b7-5e90-40eb-b8fd-e373e6568fed.png)
